### ISSUE TITLE
Windows support (msi)

### ISSUE THREAD

**Issue by GeekOffTheStreet** [jordansissel/fpm#423]
It would be cool if fpm also supported creating packages in Windows installer format.

**Comment by tartale** [comment]
+1

**Comment by rfleschenberg** [comment]
+1 ;)

**Comment by djhaskin987** [comment]
I have had some experience with windows installers since my previous comment. There are a variety of issues to deal with.

1. MSI doesn't use a the name of a package to determine whether the same package is installed or not, like RPM or DEB (or most other linux packaging systems) do. Rather, it uses a GUID specified by the developer of the MSI itself called an Upgrade Code. This can be generated from the name, so that would not be a problem for fpm, but in windows land people expect to be able to change the name, while still calling it an 'upgrade'.
2. Windows installer has no notion of `--depends`, `--provides`, `--conflicts`, etc. Those options would be moot.
3. While talking about fpm command line options, `--prefix` would make little sense. When it comes to where you install stuff, windows doesn't have much wiggle room.
4. The `--config-files` option would also make no sense. MSI does not mark certain files as config files (indeed, microsoft has no notion of manually editable configuration files).
5. The only use case that makes sense is turning a zip file or directory to an MSI. While this use case falls under the cause of FPM, adding it as an output target would mean that you could also (try to) convert an RPM or DEB to an MSI, a course of action that will almost certainly lead to disaster. The user would simply have to "know what they are doing" here.
6. On a related note, [windows has its own specific locations for where files go](https://msdn.microsoft.com/en-us/library/aa370905%28v=vs.85%29.aspx#system_folder_properties), which have no direct mapping to the linux (*nix) FHS. Further, windows assumes that the application will use the registry to keep track of stuff, a concept that is simply absent in *nix installers.

In short, I feel that FPM as currently coded is simply not the tool for the job. It makes assumptions (good assumptions!) that make it useful for packaging unix packages, but which do not help the developer at all in making a suitable, semi-useful installer.

That is not to say that I don't think making such a tool would not be useful. Perhaps a separate entrypoint could be added to the FPM codebase that is designed specifically for windows installers; the underlying codebase is built really well for making installers. But currently the entrypoint assumes unix installers, and the differences are big enough that I think it may be a mistake bolting MSI on to fpm. 

Thoughts? @jordansissel ? @hatt ?

**Comment by hatt** [comment]
Thanks for the analysis @djhaskin987, I'm actually on an MSI project atm where we shell out to templated Wix which isn't great. I know there is a ruby-ole gem that basically lets you read/write the MSI format tables. I see a use case for things like python or asset installations rather than full applications, or exploding ZIPs into installers when InnoSetup isn't a good option. Curious what Jordan thinks though since I'm just "some guy".

I do think it's possible to build MSI packages with FPM for raw directory structures but it'd essentially be limited to "explode this zip/tar archive into the prefixed directory (eg C:\Program Files (x86)\) and register the product and upgrade GUIDs".

Another option would be to output wixproj files to allow Candle/Light to build the actual MSI on a Windows machine.

**Comment by sugengsuprayogi** [comment]
+1

**Comment by shadowbq** [comment]
So maybe running fpm in Bash on Windows might give us access to Candle/light MSI build environments. 

(https://docs.microsoft.com/en-us/windows/wsl/install-win10)
(https://docs.microsoft.com/en-us/windows/wsl/interop#run-windows-tools-from-wsl)

“Windows Subsystem for Linux” provides a Windows subsystem and Ubuntu Linux runs atop it. It is not a virtual machine or an application like Cygwin. It is complete Linux system inside Windows 10.

**Comment by Songtronix** [comment]
I currently use [cargo wix](https://github.com/volks73/cargo-wix) which kind of ease the creation of an msi installer. However there's alot of customization with the raw xml files needed if you want something more than install & run. I would love FPM to replace my current setup so I don't have to jump around to finally create .msi, debian, aur packages.

**Comment by davetapley** [comment]
Or even better, MSIX, for Windows Server support:
https://learn.microsoft.com/en-us/windows/msix/msix-server-2019

**Comment by mcandre** [comment]
The desktop operating system market consists primarily of Windows (MSI, MSIX), followed by macOS, then Linux. The artifact modes supported by fpm should reflect this.

This issue was opened as far back as 2013. Unless the documentation for this is simply missing in https://fpm.readthedocs.io/en/latest/packaging-types.html , then I'm afraid I will have to pursue other tools for Windows packaging generation for my projects.

**Comment by davetapley** [comment]
Not sure if it deserves a separate issue, but there is also [WinGet](https://learn.microsoft.com/en-us/windows/package-manager/winget/) now...
